### PR TITLE
Add e2e to verify preStop hook behaviour on completed pod

### LINKED ISSUE #55807: prestop hooks aren't called when Jobs complete successfully
**Is this a BUG REPORT or FEATURE REQUEST?**: Y

> Uncomment only one, leave it on its own line: 
>
 /kind bug
> /kind feature

**What happened**:
lifecycle hook preStop is called only when user force-stop/kill the pod/job. It is not getting called when the pod is completed successfully.

**What you expected to happen**:
preStop hook should be called whenever the pod is exited either it is with success or failure
**How to reproduce it (as minimally and precisely as possible)**:
just by following instructions from https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-details

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by globervinodhn**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR adds an E2E test to verify the preStop hook behaviour on a completed pod on issuing a DELETE API call

**Which issue(s) this PR fixes**:
Fixes #55807 

**Special notes for your reviewer**:
As per the current behaviour, the preStop hook is not called for a pod in a completed state. Once the pod executes its main tasks/commands, it moves to a completed state. When a DELETE API call is issued on such pods, the pod is deleted at the earliest and the preStop is not called.

**Does this PR introduce a user-facing change?**:
NONE
release-note-none
 cc @brahmaroutu @mgdevstack

### CODE REVIEW

File: test/e2e/node/pre_stop.go [hunk 1]
```diff
@@ -219,6 +219,46 @@ var _ = SIGDescribe("PreStop", func() {
...
+	Usecase: PreStop hook is not executed for completed/terminated pod on issuing a DELETE API call
+	Description: SIGTERM and PreStop hook calls are made simultaneosly once a DELETE API call is made. However, if the main command finishes
+					execution before a DELETE API call then the pod enters the completed state. A DELETE API call would then kill and
+					remove the pod from the namespace without calling the preStop hook. Under this situation, the FailedPreStopHook
+					warnign event too would be reflecting false indicating no error.
+	Approach: 1. Sleep till the pod completes it main command and enters completed state
+			  2. Verify that the pod is alive
+			  3. Issue a DELETE API call
+			  4. Poll immediately to check if the pod is alive. If it is destroyed then preStop wasn't executed
+	*/
```
Comment (k-toyoda-pi) [inline test/e2e/node/pre_stop.go]: nits:
- warnign -> warning
- Unnecessary tab spaces are contained at the beginning of lines

Also, try "go fmt" for formarting your code.

Comment (globervinodhn) [reply test/e2e/node/pre_stop.go]: Corrected the 'warning'. I am unable to understand what TAB spaces are being referred to here. Kindly guide

**Review by pontiyaraja — CHANGES_REQUESTED** [review]
This Get() API will not return the pod which is executing the preStop process. Pod will be removed from the pod resource list map, when the delete API get called.

**Comment by globervinodhn** [comment]
> This Get() API will not return the pod which is executing the preStop process. Pod will be removed from the pod resource list map, when the delete API get called.

If you're referring to the change - "client, err = f.PodClient().Get(pod.Name, metav1.GetOptions{})" - then I do not understand why the intended pod is not returned. I have specified the name of the pod to be returned and if in case of nonavailability of the pod the concerned line will throw an error as per my understanding.

**Comment by globervinodhn** [comment]
This PR is a part of #73780, to verify the behaviour of PreStop hook

**Comment by globervinodhn** [comment]
This E2E addition can now be ignored as relevant changes to the documentation has been made via https://github.com/kubernetes/website/pull/12498.

Closing the PR.

File: test/e2e/node/pre_stop.go [hunk 2]
```diff
@@ -243,3 +283,28 @@ func getPodWithpreStopLifeCycle(name string) *v1.Pod {
...
+func getPodWithMainCmdAndPreStopLifeCycle(name string) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1.PodSpec{
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				{
+					Name:  "nginx",
```
Comment (johnbelamaric) [inline test/e2e/node/pre_stop.go]: You should probably use agnhost here instead.

**Comment by johnbelamaric** [comment]
@globervinodhn this also cannot use bare `It`.

**Comment by pontiyaraja** [comment]
@globervinodhn Please check your CLA

**Comment by globervinodhn** [comment]
> @globervinodhn Please check your CLA

I have signed it already as an individual. I am unable to figure out what exactly is the issue still

**Comment by spiffxp** [comment]
@globervinodhn please reach out to the LF helpdesk to resolve your CLA issues, https://support.linuxfoundation.org/

**Comment by spiffxp** [comment]
@globervinodhn have you had any luck? I am inclined to close this if we're not going to get anywhere with it

**Comment by spiffxp** [comment]
/close
Closing due to inactivity

**Pull request closed without merging**